### PR TITLE
Add SELinux policy for cockpit

### PR DESCRIPTION
**Pull request by dsugar100**
This was started from #700 and tested on RHEL9 system.
It is a bunch of commits because I wanted to be clear about some of the denials I was seeing and why those rules were added.
I of course will squash before the final merge (if desired).

Please let me know questions and changes desired.

### CODE REVIEW

File: policy/modules/services/cockpit.if [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+## <summary>
+##	Allow specified domain to get status of cockpit service
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`cockpit_status',`
```
Comment (pebenito) [inline policy/modules/services/cockpit.if:102]: `cockpit_get_service_status`

File: policy/modules/services/cockpit.if [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+## </param>
+#
+interface(`cockpit_startstop',`
+	gen_require(`
+		type cockpit_unit_t;
+		class service { start stop };
+	')
+
+	allow $1 cockpit_unit_t:service { start stop };
+')
```
Comment (pebenito) [inline policy/modules/services/cockpit.if:147]: `startstop` and `enabledisable` should have the `status` perm too.

File: policy/modules/services/cockpit.if [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+## <summary>
+##	Create, read, write, and delete the cockpick runtime symlink files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`cockpit_manage_runtime_lnk_files',`
```
Comment (pebenito) [inline policy/modules/services/cockpit.if:159]: `lnk_files` -> `symlinks`

File: policy/modules/services/cockpit.if [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+	gen_require(`
+		type cockpit_runtime_t;
+	')
+
+	files_search_runtime($1)
+	manage_lnk_files_pattern($1, cockpit_runtime_t, cockpit_runtime_t)
+')
+
+## <summary>
```
Comment (pebenito) [inline policy/modules/services/cockpit.if:168]: Incomplete comment block

File: policy/modules/services/cockpit.if [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+#
+interface(`cockpit_rw_session_pipes',`
+	gen_require(`
+		type cockpit_session_t;
+	')
+
+	allow $1 cockpit_session_t:fifo_file rw_fifo_file_perms;
+')
+
+## <summary>
```
Comment (pebenito) [inline policy/modules/services/cockpit.if:203]: Incomplete comment block.

File: policy/modules/services/cockpit.if [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+##	Read and write cockpit web service stream socket
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`cockpit_rw_ws_stream_socket',`
```
Comment (pebenito) [inline policy/modules/services/cockpit.if:230]: Plural "sockets"

File: policy/modules/services/cockpit.te [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+init_daemon_runtime_file(cockpit_runtime_t, lnk_file, "motd")
+
+optional_policy(`
+	systemd_tmpfilesd_managed(cockpit_runtime_t)
+')
+
+type cockpit_unit_t;
+init_unit_file(cockpit_unit_t)
+
+type cockpit_var_lib_t;
```
Comment (pebenito) [inline policy/modules/services/cockpit.te:33]: Please rename to `cockpit_state_t`.

File: policy/modules/services/cockpit.te [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+# cockpit-session communicates back with cockpit-ws
+allow cockpit_session_t cockpit_ws_t:unix_stream_socket rw_stream_socket_perms;
+
+# cockpit-tls and cockpit-ws communicate over a Unix socket
+allow cockpit_ws_t self:unix_stream_socket { create_stream_socket_perms connectto };
+
+allow cockpit_ws_t cockpit_cert_t:file unlink;
+
+ifdef(`init_systemd',`
+	systemd_exec_systemctl(cockpit_ws_t)
```
Comment (pebenito) [inline policy/modules/services/cockpit.te:134]: This can be a regular optional.

File: policy/modules/services/cockpit.te [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+dbus_system_bus_client(cockpit_session_t)
+
+miscfiles_read_localization(cockpit_session_t)
+
+ifdef(`init_systemd',`
+	systemd_dbus_chat_logind(cockpit_session_t)
+	systemd_search_all_user_keys(cockpit_session_t)
+	systemd_write_all_user_keys(cockpit_session_t)
+')
```
Comment (pebenito) [inline policy/modules/services/cockpit.te:207]: Can be an optional block.

File: policy/modules/system/init.if [hunk 10]
```diff
@@ -3044,6 +3044,24 @@ interface(`init_manage_utmp',`
...
+## <summary>
+##	Add a watch on init runtime
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`init_watch_runtime',`
```
Comment (pebenito) [inline policy/modules/system/init.if:3057]: `init_watch_runtime_dirs`

File: policy/modules/system/init.te [hunk 11]
```diff
@@ -303,6 +303,7 @@ ifdef(`init_systemd',`
 
 	allow init_t init_runtime_t:sock_file manage_sock_file_perms;
 
+	allow init_t daemon:process noatsecure;
```
Comment (pebenito) [inline policy/modules/system/init.te:306]: This should be a separate patch. Also it wouldn't be here, but rather in `init_domain`, `init_daemon_domain` etc.

File: policy/modules/system/systemd.if [hunk 12]
```diff
@@ -161,6 +162,8 @@ template(`systemd_role_template',`
 
 	can_exec($3, systemd_analyze_exec_t)
 
+	domtrans_pattern($1_systemd_t, systemd_tmpfiles_exec_t, systemd_tmpfiles_t)
```
Comment (pebenito) [inline policy/modules/system/systemd.if:165]: The systemd user sessions definitely shouldn't be starting a system service.

**Review by pebenito — CHANGES_REQUESTED** [review]
`cockpit.te` also needs some style cleanups.

File: policy/modules/system/systemd.if (hunk 12 above)
Comment (dsugar100) [reply policy/modules/system/systemd.if:165]: I'm open to suggestions on this one.  I started down the path of setting up a user tmpfiles concept (i.e. setting up $1_systemd_tmpfiles_t), but it quickly got out of hand on permissions.  I can try this approach again unless there is a better alternative?

File: policy/modules/services/cockpit.te (hunk 9 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.te:207]: updated

File: policy/modules/system/init.if (hunk 10 above)
Comment (dsugar100) [reply policy/modules/system/init.if:3057]: fixed

File: policy/modules/system/systemd.if (hunk 12 above)
Comment (0xC0ncord) [reply policy/modules/system/systemd.if:165]: I agree transitioning from the systemd user session to the tmpfiles domain directly is a pretty significant elevation of privileges.

What kinds of things does this operation need to do? I'm not familiar with the innerworkings of cockpit and how this is related.

File: policy/modules/services/cockpit.if (hunk 2 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.if:147]: done - added use of 'cockpit_status'

File: policy/modules/services/cockpit.if (hunk 1 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.if:102]: renamed

File: policy/modules/services/cockpit.if (hunk 3 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.if:159]: renamed

File: policy/modules/services/cockpit.if (hunk 6 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.if:230]: renamed

File: policy/modules/system/init.te (hunk 11 above)
Comment (dsugar100) [reply policy/modules/system/init.te:306]: Moved into init_daemon_domain and into patch #752

File: policy/modules/services/cockpit.fc [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+/usr/lib/systemd/system/cockpit.*		--	gen_context(system_u:object_r:cockpit_unit_t,s0)
+
+/usr/libexec/cockpit-certificate-ensure	--	gen_context(system_u:object_r:cockpit_cert_manage_exec_t,s0)
+/usr/libexec/cockpit-ws		--	gen_context(system_u:object_r:cockpit_ws_exec_t,s0)
+/usr/libexec/cockpit-tls	--	gen_context(system_u:object_r:cockpit_ws_exec_t,s0)
+/usr/libexec/cockpit-wsinstance-factory	--	gen_context(system_u:object_r:cockpit_ws_exec_t,s0)
+/usr/libexec/cockpit-session	--	gen_context(system_u:object_r:cockpit_session_exec_t,s0)
+/usr/libexec/cockpit-ssh	--	gen_context(system_u:object_r:cockpit_session_exec_t,s0)
+
+/usr/share/cockpit/motd/update-motd    -- gen_context(system_u:object_r:cockpit_ws_exec_t,s0)
```
[the remark this answers is not in the document]
Comment (dsugar100) [reply policy/modules/services/cockpit.fc:17]: Either because I don't think we wanted bin_t in cockpit.fc and/or because I didn't think we want /usr/share/cockpit/... in corecommands.fc (which upon further inspection might be OK) and/or because as I was working on this to get things running in enforcing I ended up with this and didn't give it much more thought.  

Either way, bin_t seems like a reasonable label.  I will do more testing to ensure it works as expected.

File: policy/modules/services/cockpit.if (hunk 5 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.if:203]: It too me a stupid amount of time to realize it was missing the line of '#'s above 'summary'.

File: policy/modules/services/cockpit.if (hunk 4 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.if:168]: done

**Comment by dsugar100** [comment]
Changed status to 'draft'.   I'm still working through changes.  I'm happy for more feedback, but don't want anyone to think my updates are indications that I'm done.

File: policy/modules/admin/sudo.if [hunk 14]
```diff
@@ -203,6 +204,8 @@ template(`sudo_role_template',`
 
 		ifdef(`init_systemd',`
 			init_dbus_chat($1_sudo_t)
+			systemd_dbus_chat_logind($1_sudo_t)
+			systemd_read_logind_state($1_sudo_t)
 		')
 	')
```
[the remark this answers is not in the document]
Comment (dsugar100) [reply policy/modules/admin/sudo.if]: I think I have updated as you asked.  I did leave the "ifdef(init_systemd)" as that was there before.  Let me know if I misunderstood or you want additional changes.

File: policy/modules/services/ntp.te [hunk 15]
```diff
@@ -171,6 +171,7 @@ ifdef(`init_systemd',`
 	optional_policy(`
 		chronyd_enabledisable(ntpd_t)
 		chronyd_startstop(ntpd_t)
+		chronyd_status(ntpd_t)
```
Comment (dsugar100) [inline policy/modules/services/ntp.te:174]: Related to the comment about cockpit_enabledisable and cockpit_startstop granting 'status' access, should this be here, or should I update chronyd_enabledisable and chronyd_startstop to allow 'status' access instead?

File: policy/modules/services/cockpit.if [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+##	<summary>
+##	Role allowed access
+##	</summary>
+## </param>
+#
+template(`cockpit_role_template',`
+
+	type $1_cockpit_memfd_t;
+	files_runtime_file($1_cockpit_memfd_t)
+	dev_filetrans($2, $1_cockpit_memfd_t, file)
```
[the remark this answers is not in the document]
Comment (dsugar100) [reply policy/modules/services/cockpit.if:50]: It maybe that the name is just wrong there.  But, there is access where cockpit-bridge (running as the user) is attempting to create and file and then execute that file.  The default type was user_tmpfs_t (and I didn't want to grant blanket execute permission on that type).  It does try creating these files in several different locations.  It seemed to be part of libffi (foreign-function interface).  

Maybe $1_cockpit_tmpfs_t would be a better type?

File: policy/modules/system/systemd.if (hunk 12 above)
Comment (dsugar100) [reply policy/modules/system/systemd.if:165]: Ok, after reviewing this, it does seem that $1_systemd_tmpfiles_t is working ok and not an unreasonable amount of policy.  Maybe I was looking at this too soon during the development of the overall cockpit poilcy and was seeing something else that was confusing me.

**Comment by dsugar100** [comment]
Ok, I have made updates.  There is a remaining question about the memfd thing which I will update after feedback and new question about chronyd_status.  Please let me know if there is anything else.

File: policy/modules/admin/sudo.if (hunk 14 above)
[the remark this answers is not in the document]
Comment (pebenito) [reply policy/modules/admin/sudo.if]: Ideally everything for enabling systemd would depend on the presence of the systemd module. In cases like this, the `ifdef(init_systemd)` is because there wasn't anything depending on the systemd module. By adding the `systemd_*` calls, now there are dependencies.

File: policy/modules/services/cockpit.if (hunk 16 above)
[the remark this answers is not in the document]
Comment (pebenito) [reply policy/modules/services/cockpit.if:50]: Lets go with the `tmpfs` label. By default memfd_create() will result in new shm segments. shm segments should be `tmpfs` in other places in the policy.

For future reference, here's the set of locations libffi will try to create the write+exec file (at the time I looked for this comment): https://github.com/libffi/libffi/blob/622caabcd25f4e11f752241417d06c9062acdf1f/src/closures.c#L726

[the remark this answers is not in the document]
Comment (dsugar100) [reply policy/modules/services/cockpit.if]: Ok, updated to $1_cockpit_tmpfs_t

File: policy/modules/admin/sudo.if (hunk 14 above)
[the remark this answers is not in the document]
Comment (dsugar100) [reply policy/modules/admin/sudo.if:206]: Ok, ifdef(init_systemd) also removed.

**Comment by dsugar100** [comment]
Ok, I believe at this point all your comments have been addressed.   Do you want me to resolve conversations that I believe are complete?
Thanks for the time, I know this is a big pull request.

File: policy/modules/admin/sudo.if (hunk 14 above)
[the remark this answers is not in the document]
Comment (pebenito) [reply policy/modules/admin/sudo.if:206]: Sorry, I made a mistake. Please keep the ifdef. I confused this with one of the other changes.

File: policy/modules/services/ntp.te (hunk 15 above)
Comment (pebenito) [reply policy/modules/services/ntp.te:174]: > should I update chronyd_enabledisable and chronyd_startstop to allow 'status' access instead?

Yes, this.

File: policy/modules/admin/sudo.if (hunk 14 above)
[the remark this answers is not in the document]
Comment (dsugar100) [reply policy/modules/admin/sudo.if:206]: replace ifdef

File: policy/modules/services/cockpit.te (hunk 8 above)
Comment (dsugar100) [reply policy/modules/services/cockpit.te:132]: moved

File: policy/modules/services/ntp.te (hunk 15 above)
Comment (dsugar100) [reply policy/modules/services/ntp.te:174]: updated

**Comment by dsugar100** [comment]
Ok, I have updated based on the three comments about the ifdef, chronyd_status and reordering.  Let me know if anything else is needed.

**Review by pebenito — APPROVED** [review]
